CRM/Case/BAO/Case.php

```diff
@@ -1309,7 +1309,6 @@ public static function sendActivityCopy($clientId, $activityId, $contacts, $atta
 
       list($result[CRM_Utils_Array::value('contact_id', $info)], $subject, $message, $html) = CRM_Core_BAO_MessageTemplate::sendTemplate(
         [
-          'groupName' => 'msg_tpl_workflow_case',
           'valueName' => 'case_activity',
```

Comment:
So, just verbalizing a bit, this PR basically flattens the namespace - e.g. to identify a template, you previously needed the tuple `(groupName,valueName)` (aka `(civicrm_option_group.name,civicrm_option_value.name)`) - but now it only requires `(valueName)`.

The implication is that `groupName` was superfluous - and, at least as far as the default data in `civicrm-core` goes, this appears to check out. The `civicrm_option_value.name`s are unique:

```
mysql> select id, name, label, value 
  from civicrm_option_value 
  where option_group_id in (select id from civicrm_option_group where name like 'msg_tpl_workflo%') 
  order by name;

+-----+----------------------------------+----------------------------------------------------------------+-------+
| id  | name                             | label                                                          | value |
+-----+----------------------------------+----------------------------------------------------------------+-------+
| 820 | case_activity                    | Cases - Send Copy of an Activity                               | 1     |
| 821 | contribution_dupalert            | Contributions - Duplicate Organization Alert                   | 1     |
| 824 | contribution_invoice_receipt     | Contributions - Invoice                                        | 4     |
| 822 | contribution_offline_receipt     | Contributions - Receipt (off-line)                             | 2     |
| 823 | contribution_online_receipt      | Contributions - Receipt (on-line)                              | 3     |
| 827 | contribution_recurring_billing   | Contributions - Recurring Billing Updates                      | 7     |
| 826 | contribution_recurring_cancelled | Contributions - Recurring Cancellation Notification            | 6     |
| 828 | contribution_recurring_edit      | Contributions - Recurring Updates                              | 8     |
| 825 | contribution_recurring_notify    | Contributions - Recurring Start and End Notification           | 5     |
| 834 | event_offline_receipt            | Events - Registration Confirmation and Receipt (off-line)      | 1     |
| 835 | event_online_receipt             | Events - Registration Confirmation and Receipt (on-line)       | 2     |
| 836 | event_registration_receipt       | Events - Receipt only                                          | 3     |
| 841 | friend                           | Tell-a-Friend Email                                            | 1     |
| 845 | membership_autorenew_billing     | Memberships - Auto-renew Billing Updates                       | 4     |
| 844 | membership_autorenew_cancelled   | Memberships - Auto-renew Cancellation Notification             | 3     |
| 842 | membership_offline_receipt       | Memberships - Signup and Renewal Receipts (off-line)           | 1     |
| 843 | membership_online_receipt        | Memberships - Receipt (on-line)                                | 2     |
| 837 | participant_cancelled            | Events - Registration Cancellation Notice                      | 4     |
| 838 | participant_confirm              | Events - Registration Confirmation Invite                      | 5     |
| 839 | participant_expired              | Events - Pending Registration Expiration Notice                | 6     |
| 840 | participant_transferred          | Events - Registration Transferred Notice                       | 7     |
| 833 | payment_or_refund_notification   | Additional Payment Receipt or Refund Notification              | 13    |
| 829 | pcp_notify                       | Personal Campaign Pages - Admin Notification                   | 9     |
| 832 | pcp_owner_notify                 | Personal Campaign Pages - Owner Notification                   | 12    |
| 830 | pcp_status_change                | Personal Campaign Pages - Supporter Status Change Notification | 10    |
| 831 | pcp_supporter_notify             | Personal Campaign Pages - Supporter Welcome                    | 11    |
| 851 | petition_confirmation_needed     | Petition - need verification                                   | 2     |
| 850 | petition_sign                    | Petition - signature added                                     | 1     |
| 847 | pledge_acknowledge               | Pledges - Acknowledgement                                      | 1     |
| 848 | pledge_reminder                  | Pledges - Payment Reminder                                     | 2     |
| 846 | test_preview                     | Test-drive - Receipt Header                                    | 1     |
| 849 | uf_notify                        | Profiles - Admin Notification                                  | 1     |
| 899 | volunteer_registration           | Volunteer - Registration (on-line)                             | 1     |
+-----+----------------------------------+----------------------------------------------------------------+-------+
```

So this line (like several of the others) is demonstrating that `CRM_Core_BAO_MessageTemplate::sendTemplate()` no longer requires the `groupName`. But is it *harmful* to pass it in anyway?

The reason I ask - in `universe`, I found 9 extensions which call `MessageTemplate::sendTemplate()` -- `au.com.agileware.ewayrecurring`, `biz.jmaconsulting.bugp`, `com.agiliway.civimobileapi`, `com.chrischinchilla.ewayrecurring`, `org.civicrm.cdntaxreceipts`,
`org.civicrm.volunteer`, `uk.co.compucorp.civicrm.booking`, `uk.co.vedaconsulting.civisubscription`, `uk.co.vedaconsulting.gdpr`. I assume that there'd be some mix of deployments/configurations in which (for the near term) most would still be passing it in.

Additionally, you mentioned this in the description:

> Also note this DOES mean that we are no longer passing the optionGroup to the hook. I will email the dev list on that

If we downgrade `groupName` from *required* to *optional*, then perhaps it's a smoother transition?

Reply:
@totten so the plan was to follow up with a patch that merges all the option groups into one and then we can update workflow_id to hold the value not the id & we have a normal option group & we can stop working around this weird structure.

Those extensions you mention won't be affected - the groupName parameter will be ignored. However, they all pass a group name with the msg_tpl we are looking for, so they will all still work - eg  https://github.com/jake-mw/CDNTaxReceipts/blob/4df12a4c501c556bf90947c49875b9a1e1ac682e/cdntaxreceipts.functions.inc#L125

So the change would only affect extensions which listen to the hook and specifically pay attention to the group name - hence my email heads up to the dev list.

However, the expectation is that groupName will not stay the same going forwards anyway. We would do something like 
 add a new core option group 'msg_tpl_workflow' and migrate existing groups into it so we only have one option group, but keep it working off the option_group.name LIKE for w while, adding some deprecation noise over time to push extensions to switch over.

I think we would do that fairly well in advance of updating workflow_id to hold the integer from 'value' rather than 'id' and make sure we have api support in place so that no-one should care whether workflow_id is a weird option value - because they can use the api to interact with it.